pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go

```diff
@@ -132,15 +141,38 @@ func (og *operationGenerator) GenerateRegisterPluginFunc(
 	return registerPluginFunc
 }
 
+func unregisterPlugin(actualStateOfWorldUpdater ActualStateOfWorldUpdater, socketPath string, handler cache.PluginHandler, name string) {
+	// We remove the plugin to the actual state of world cache before calling a plugin consumer's Unregister handle
+	// so that if we receive a register event during Register Plugin, we can process it as a Register call.
+	actualStateOfWorldUpdater.RemovePlugin(socketPath)
+
+	handler.DeRegisterPlugin(name)
+}
 func (og *operationGenerator) GenerateUnregisterPluginFunc(
 	socketPath string,
 	pluginHandlers map[string]cache.PluginHandler,
+	pathToHandlers *cache.SocketPluginHandlers,
 	actualStateOfWorldUpdater ActualStateOfWorldUpdater) func() error {
 
 	unregisterPluginFunc := func() error {
 		client, conn, err := dial(socketPath, dialTimeoutDuration)
 		if err != nil {
-			return fmt.Errorf("UnregisterPlugin error -- dial failed at socket %s, err: %v", socketPath, err)
+			klog.V(4).Infof("unable to dial: %v", err)
+			var handlerWithName cache.NamedPluginHandler
+			handlerFound := false
+			pathToHandlers.Lock()
+			handlerWithName, handlerFound = pathToHandlers.Handlers[socketPath]
+			pathToHandlers.Unlock()
+
+			if !handlerFound {
+				return fmt.Errorf("unable to DeRegister - there is no handler for %s", socketPath)
+			}
+			unregisterPlugin(actualStateOfWorldUpdater, socketPath, handlerWithName.Handler, handlerWithName.Name)
+			pathToHandlers.Lock()
+			delete(pathToHandlers.Handlers, socketPath)
+			pathToHandlers.Unlock()
+			klog.V(4).Infof("DeRegisterPlugin called for %s on %v", handlerWithName.Name, handlerWithName.Handler)
+			return nil
 		}
```

Comment:
I think we should only keep the `dial`, but remove `client.GetInfo`. This way, we can move everything after `var handlerWithName cache.NamedPluginHandler` out of the if statement, and have one path for unregistration.

Reply:
@taragu 
Can you confirm that we are abandoning the old code (except for dial) and would use the new code ?

Reply:
@taragu 
Can you take another look ?

Reply:
@taragu
Please review

Reply:
@taragu 
Please take another look